Ethereal stomach damage and failure, other qol and bugfixes

KittyNoodle opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Ethereal charge alert buxfix. Fully upgraded chargers will no longer overcharge ethereals. Makes overcharge more dangerous with the new stomach damage and failure states, as overcharge will damage the stomach(empty battery will also very slowly damage the stomach). Damaged ethereal stomach will discharge sooner. Failed ethereal stomach will jump to zero charge and start killing the ethereal.

## Why It's Good For The Game
closes #61063. Makes being an ethereal less painful and overcharge less often. Having more fun failure states is good.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Fail and damage states to biological battery
qol: chargers with femto manipulators will not overcharge ethereals
fix: ethereal charge alerts will no longer stay on screen when outdated
balance: ethereal overcharge no longer has a chance to cause heart attack, will cause the stomach to fail instead.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

KittyNoodle committed:
Unique damage text

KittyNoodle committed:
Damage effects added

KittyNoodle committed:
Overcharge damages the battery

KittyNoodle committed:
Damage cap because capping damage is good

KittyNoodle committed:
Adds failure response

KittyNoodle committed:
bugfix

KittyNoodle committed:
Update stomach_ethereal.dm

KittyNoodle committed:
more bugfix

KittyNoodle committed:
bugfixes, spellchecks, and rebalances

KittyNoodle committed:
tweaking things

KittyNoodle committed:
borg chargers with repair no longer overcharge ethereals

KittyNoodle committed:
consistency in definitions

KittyNoodle committed:
i did something dumb

KittyNoodle committed:
alert fix(sloppy edition)

ATH1909 commented on code/modules/surgery/organs/stomach/stomach_ethereal.dm:
```diff
@@ -6,11 +6,34 @@
+	var/damagemodifer = 0.5
+	low_threshold_passed = "<span class='info'>You can feel your power draining as a feeling of pins and needles fills your abdomen.</span>"
+	high_threshold_passed = "<span class='warning'>Your abdomen is filled with the feeling of shocks and electric discharges!</span>"
+	high_threshold_cleared = "<span class='info'>The electric feeling in your stomach seems to fade away, leaving behind a feeling of pins and needles</span>"
```
```suggestion
	high_threshold_cleared = "<span class='info'>The electric feeling in your stomach seems to fade away, leaving behind a feeling of pins and needles.</span>"
```

ATH1909 commented on code/modules/surgery/organs/stomach/stomach_ethereal.dm:
```diff
@@ -6,11 +6,34 @@
+	if(damage < low_threshold)
+		return
+	adjust_charge(-0.8 * (damage / 5) * delta_time) //at 100 damage ethereals will loose 16 charge per second
+	handle_charge(owner, delta_time, times_fired)
```
this will cause being on low charge to damage you twice as fast

KittyNoodle replied:
fixed

ATH1909 commented on code/modules/surgery/organs/stomach/stomach_ethereal.dm:
```diff
@@ -6,11 +6,34 @@
+		damage = maxHealth
+	if(damage < low_threshold)
+		return
+	adjust_charge(-0.8 * (damage / 5) * delta_time) //at 100 damage ethereals will loose 16 charge per second
```
oh god, this sound horrible to deal with. you get defibbed, then need to spend the rest of your shift next to a cyborg recharger because the stomach repair surgery can only be initiated on stomachs with >50 damage.

ATH1909 replied:
like, to put this in perspective, let's lowball the amount of charge decay from this at the lowest possible number, an extra -1.6 points of charge per delta_time. if I'm correct in assuming that 1 point of delta_time correlates to 1 second passing, this would reduce the amount of time it takes to go from ETHEREAL_CHARGE_FULL to ETHEREAL_CHARGE_NORMAL (below which you take a negative moodlet) from 20.8 minutes to **6.9 minutes**. If you use the decay rate for 20 stomach damage instead of the decay rate for 10 points of stomach damage, you'll go from full to a negative moodlet in  just **4.1 minutes**. This is assuming that you know exactly when to stop charging to prevent overcharging, btw.

If this change goes through, it will deadass become the meta for newly defibbed ethereals to immediately kill themselves to repair their stomachs.

KittyNoodle replied:
gonna make it take as long as brain then

KittyNoodle replied:
> This is assuming that you know exactly when to stop charging to prevent overcharging, btw.
> 
qol: chargers with femto manipulators will not overcharge ethereals

ATH1909 commented on code/modules/surgery/organs/stomach/stomach_ethereal.dm:
```diff
@@ -6,11 +6,34 @@
+	if(!(organ_flags & ORGAN_FAILING))
+		return
+	if(organ_owner)
+		organ_owner.adjustToxLoss(0.65 * delta_time, TRUE, TRUE)
```
this damage should be capped like the damage from having low charge is

KittyNoodle replied:
no, that is kinda the point, it is supposed to be fatal at failure

ATH1909 replied:
the damage cap for being low on charge is kind of pointless if you can die from being low on charge anyway.

also, no other species outright dies from (just) being low on hunger; why should ethereals?

KittyNoodle replied:
The uncapped damage only happens if the organ is failing, not under normal conditions

ATH1909 replied:
> The uncapped damage only happens if the organ is failing, not under normal conditions

yeah, but the organ can fail from the damage from being low on charge

KittyNoodle replied:
you need to, wait to get to empty charge, then wait another 5 minutes, i wanted ethereals to have more danger in going with zero charge for extended periods of time

KittyNoodle committed:
spellcheck

Co-authored-by: ATH1909 <42606352+ATH1909@users.noreply.github.com>

KittyNoodle committed:
makes defibbing ethereals not pain

KittyNoodle committed:
I made a stupid

KittyNoodle committed:
Vital comment update

ATH1909 commented:
reducing the decay rate of ethereal stomachs won't fix the issue here. under this system, being overcharged for >20 seconds will cause your stomach to hit that 10 damage threshold and make your life hell.

KittyNoodle commented:
stomach organ damage is kinda sharp in all versions, with regular stomach you just start vomiting all the time. and i did some testing and the decay time seems good

LemonInTheDark commented:
> Fail and damage states to biological battery

What does this mean?

KittyNoodle commented:
> > Fail and damage states to biological battery
> 
> What does this mean?

Most organs do things when they are damaged or non functional (eg, damaged lungs needing more air and broken lungs not breathing)

wesoda25 commented on code/modules/surgery/organs/stomach/stomach_ethereal.dm:
```diff
@@ -6,11 +6,35 @@
+	if(damage < low_threshold)
+		handle_charge(owner, delta_time, times_fired)
+		return
+	adjust_charge(-0.8 * (damage / 5) * delta_time) //at 100 damage ethereals will loose 16 charge per second
```
Please base this equation (and all like it) off of ETHEREAL_CHARGE_FACTOR so that it's updated accordingly should values be changed down the road.

wesoda25 commented on code/modules/surgery/organs/stomach/stomach_ethereal.dm:
```diff
@@ -50,27 +78,37 @@
+			carbon.clear_alert("ethereal_charge")
 			carbon.throw_alert("ethereal_overcharge", /atom/movable/screen/alert/ethereal_overcharge, 1)
 			carbon.apply_damage(0.2, TOX, null, null, carbon)
+			damage += (1 * damagemodifer)
```
Please either lower or remove the damage for this stage, as it is often very easy to ethereals to accidentally stray into the overload territory and I'd rather not heavily punish them for it.

KittyNoodle replied:
with the new changes to cyborg chargers and ethereals overcharge is less likely, also the variable damagemodifer is at 0.5, so this damage is 0.5 over time not 1.

AMonkeyThatCodes replied:
I still want what wesoda said to be done, because cyborg charger is not the only mean of charging.

LemonInTheDark replied:
overload is punished by shooting bolts of lightning isn't it? if you want to do this you should give more reasoning for why it should be done

KittyNoodle replied:
Critical overcharge lasts 10 seconds tops before you vomit out lightning, i have tested this before and the damage from overcharge is insignificant, note that the damage is actually 0.25 because of the damage modifier

wesoda25 replied:
> overload is punished by shooting bolts of lightning isn't it? if you want to do this you should give more reasoning for why it should be done

You're thinking of discharging, which is only possible in the CHARGE_DANGEROUS threshold. CHARGE_OVERLOAD - what we're talking about - is effectively the grace threshold between full charge and dangerous. It's relatively easy to fall into, be it from rechargers or not (although before this they were definitely the worst culprits), hence why we're trying to keep the damage for this threshold minimal.

The value they chose actually isn't too terrible, but in context with the other values for varying amounts of charge I don't think it makes sense. With the current values, an ethereal on an empty stomach will take half of much damage as one in the overload area, yet it's far easier to find yourself in overload than empty. I think the values should reflect the comparative difficulty of falling into a category. That being said, I'm definitely fine with the DANGEROUS area to deal a lot, cause it both makes sense and there are protections in place to prevent anyone from being in said category too long.

KittyNoodle replied:
the damage it does is very low and if you overcharge via apc you can just put the power back in. the times i have overcharged have been rare and far between, and it usually is dealt with within a minute, also considering that with more damage the faster charge is consumed meaning that you spend less time in overcharge, also the organ still heals during this time.

KittyNoodle committed:
Minor change for future editing

KittyNoodle committed:
Balance changes

KittyNoodle committed:
Balance changes

KittyNoodle commented:
@Mothblocks pls

AMonkeyThatCodes commented:
there's still an open review ;^)

KittyNoodle commented:
Stale moment

ExcessiveUseOfCobblestone commented:
> I thought I did

nbd just Include screenshot in future then

KittyNoodle committed:
Formatting and singular grammar fix

KittyNoodle committed:
Programming in Java has broken me

ExcessiveUseOfCobblestone commented:
your PR is getting softlocked because you havent responded https://github.com/tgstation/tgstation/pull/61142#pullrequestreview-746549085

KittyNoodle committed:
Tops off ethereal charge instead of charging them up to a specific point

Prevents a rare occurrence where if a charger was highly upgraded enough it would not charge ethereals beyond low power

KittyNoodle commented:
Rip
